history/history.c

```diff
@@ -0,0 +1,1081 @@
+/*
+ * Copyright (C) 2022 VMware, Inc. All Rights Reserved.
+ *
+ * Licensed under the GNU Lesser General Public License v2.1 (the "License");
+ * you may not use this file except in compliance with the License. The terms
+ * of the License are located in the COPYING file of this distribution.
+ */
+
+#include <stdio.h>
+#include <time.h>
+
+#include <sqlite3.h>
+
+#include <rpm/rpmlib.h>
+#include <rpm/rpmdb.h>
+#include <rpm/rpmlog.h>
+#include <rpm/rpmps.h>
+#include <rpm/rpmts.h>
+
+#include "history.h"
+
+#define check_cond(COND) if(!(COND)) { \
+    fprintf(stderr, "check_cond failed in %s line %d\n", \
+        __FUNCTION__, __LINE__); \
+    rc = -1; \
+    ((void)(rc)); /* suppress "set but not used" warning */ \
+    goto error; \
+}
+
+#define check_ptr(ptr) if(!(ptr)) { \
+    fprintf(stderr, "check_ptr failed in %s line %d\n", \
+        __FUNCTION__, __LINE__); \
+    rc = -1; \
+    ((void)(rc)); /* suppress "set but not used" warning */ \
+    goto error; \
+}
+
+#define check_rc(rc) if((rc) != 0) { \
+    fprintf(stderr, "check_rc failed in %s line %d\n", \
+        __FUNCTION__, __LINE__); \
+    goto error; \
+}
+
+#define check_db_rc(db, rc) if((rc) != SQLITE_OK) { \
+    fprintf(stderr, \
+        "check_db_rc failed with %s in %s line %d\n", \
+        sqlite3_errmsg(db), __FUNCTION__, __LINE__); \
+    ((void)(rc)); /* suppress "set but not used" warning */ \
+    goto error; \
+}
+
+#define check_db_cmd(db, cmd) { \
+    rc = (cmd); \
+    check_db_rc(db, rc); \
+}
+
+#define safe_free(ptr) { if ((ptr) != NULL) { free(ptr); ptr = NULL; }}
+
+#define map_isset(map, idx) (map[(idx)-1] != 0)
+#define map_set(map, idx) { map[(idx)-1] = 1;}
+#define map_unset(map, idx) { map[(idx)-1] = 0;}
+
+static
+int _cmp_int(const void *p1, const void *p2)
+{
+    return *(int *)p1 - *(int *)p2;
+}
+
+static
+void sort_array(int *arr, int count)
+{
+    qsort(arr, count, sizeof(int), _cmp_int);
+}
+
+/*
+ * Diff arr1 and arr2, with sizes count1 and count2. The arrays must be
```

Comment:
We can simplify this further:
```
static
int diff_arrays(int *arr1, int count1,
                 int *arr2, int count2,
                 int **ponly1, int *ponly1_count,
                 int **ponly2, int *ponly2_count)
{
    int rc = 0;
    int i1 = 0, i2 = 0;
    int *only1 = NULL, *only2 = NULL;
    int only1_count = 0, only2_count = 0;

    check_ptr(only1 = (int *)calloc(count1, sizeof(int)));
    check_ptr(only2 = (int *)calloc(count2, sizeof(int)));

    for (; i1 < count1 && i2 < count2; i1++, i2++) {
        if (arr1[i1] != arr2[i2]) {
            only1[only1_count++] = arr1[i1];
            only2[only2_count++] = arr2[i2];
        }
    }

    while (i2 < count2) {
        only2[only2_count++] = arr2[i2++];
    }

    while (i1 < count1) {
        only1[only1_count++] = arr1[i1++];
    }

    *ponly1 = only1;
    *ponly1_count = only1_count;

    *ponly2 = only2;
    *ponly2_count = only2_count;

error:
    if(rc) {
        safe_free(only1);
        safe_free(only2);
    }
    return rc;
}
```

Reply:
I don't think that works. If `arr1` is` [1, 2]` and `arr2` is `[2, 3]`, we'd add `2` to `only2` on the first iteration of the loop, and that's wrong.